packages/frontend/editor-ui/src/features/ai/assistant/builder.store.ts

```diff
@@ -776,6 +852,70 @@ export const useBuilderStore = defineStore(STORES.BUILDER, () => {
 		{ deep: true },
 	);
 
+	// Version management for workflow history
+	const workflowHistoryStore = useWorkflowHistoryStore();
+
+	/**
+	 * Restores the workflow to a previous version and truncates chat messages.
+	 * Finds the user message with the matching messageId and removes it
+	 * along with all messages after it.
+	 *
+	 * @param versionId - The workflow version ID to restore to
+	 * @param messageId - The message ID to truncate from
+	 */
+	async function restoreToVersion(
+		versionId: string,
+		messageId: string,
+	): Promise<IWorkflowDb | undefined> {
+		const workflowId = workflowsStore.workflowId;
+
+		// Save current workflow if there are unsaved changes before restoring
+		if (uiStore.stateIsDirty) {
+			const saved = await workflowSaver.saveCurrentWorkflow();
+			if (!saved) {
+				// saving errored or user opted not to overwrite changes
+				return;
+			}
+		}
+
+		// 1. Restore the workflow using existing workflow history store
+		const updatedWorkflow = await workflowHistoryStore.restoreWorkflow(
+			workflowId,
+			versionId,
+			false,
+		);
+
+		// version id is important to update, because otherwise the next time user saves,
+		// "overwrite" prevention modal shows, because the version id on the FE would be out of sync with latest on the backend
+		workflowState.setWorkflowProperty('versionId', updatedWorkflow.versionId);
+		workflowState.setWorkflowProperty('updatedAt', updatedWorkflow.updatedAt);
+
+		// 2. Truncate messages in backend session (removes message with messageId and all after)
+		await truncateBuilderMessages(rootStore.restApiContext, workflowId, messageId);
```

Comment:
One thing I noted using this locally, the truncation happening after the restore looks a little odd, it might make sense to do both the truncation and restoring of the workflow at the same time with `Promise.all([...])`

Reply:
Ya maybe UX could be improved. But if restore API request fails for some reason, I think it would be better to keep the context.

Reply:
Maybe a loading state is needed. Or we fade the messages or something. 🤔 I would say out of scope for now though.